[EPMEDU-3098]: Implement feedings page

yuriiq opened the pull request:
[Jira ticket](https://jira.epam.com/jira/browse/EPMEDU-3098)

### Description
- Implemented displaying recent feedings (less than 10 days ago) with `pending`, `approved`, `rejected` and `outdated` statuses

### Video
https://github.com/AnimealProject/animeal_android/assets/83027107/32d032ad-84ae-48bc-817f-dcaf9fc12e14

P.S. Other statuses and viewing as moderator will be implemented in separate PRs.
P.P.S. Subscriptions on feedings updates are not working at the moment. Reported to the backend team.

yuriiq committed:
EPMEDU-1506: add feedings to the menu

yuriiq committed:
[EPMEDU-1506]: create feedings list

yuriiq committed:
[EPMEDU-1506]: add pending statuses

yuriiq committed:
[EPMEDU-1506]: add GetFeedingPointsUseCase with status

yuriiq committed:
[EPMEDU-1506]: codestyle fix

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/model/FeedingHistory.kt:
```diff
@@ -6,5 +6,6 @@ data class FeedingHistory(
     val id: String,
     val name: String,
     val surname: String,
+    val status: type.FeedingStatus,
```
Please create a separate enum in this package to use here. Domain layer must not know about data (generated models) nor presentation layers

yuriiq replied:
fixed

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/usecase/GetFeedingHistoriesUseCase.kt:
```diff
@@ -8,8 +8,12 @@ import kotlinx.coroutines.flow.map
 class GetFeedingHistoriesUseCase(private val repository: FeedingRepository) {
```
Let's rename this usecase to `GetApprovedFeedingHistoriesUseCase` to avoid confusion

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/di/FeedingHistoriesModule.kt:
```diff
@@ -0,0 +1,21 @@
+
+@Module
+@InstallIn(ViewModelComponent::class)
+internal object FeedingHistoriesModule {
```
No need to create a module for each usecase, just put both providers in one module

yuriiq replied:
fixed

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/domain/usecase/GetFeedingPointsUseCase.kt:
```diff
@@ -0,0 +1,8 @@
+
+import com.epmedu.animeal.feeding.domain.repository.FeedingPointRepository
+
+class GetFeedingPointsUseCase(private val feedingPointRepository: FeedingPointRepository) {
```
Let's rename `GetAllFeedingPointsUseCase` to `GetFeedingPointsByTypeUseCase` to avoid confusion

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedStatusItemUI.kt:
```diff
@@ -0,0 +1,68 @@
+            )
+            Divider()
+            FeedStatusItem(
+                status = FeedStatus.REJECTED
```
Can you please add all statuses to preview?

yuriiq replied:
added

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+@OptIn(ExperimentalMaterialApi::class)
+@Suppress("LongMethod")
+@Composable
+fun FeedingApproveItem(
```
You forgot to rename this function :)

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+import com.epmedu.animeal.foundation.theme.AnimealTheme
+import com.epmedu.animeal.foundation.theme.CustomColor
+
+@OptIn(ExperimentalMaterialApi::class)
```
Use a `Card` constructor that is not experimental

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+import com.epmedu.animeal.foundation.theme.CustomColor
+
+@OptIn(ExperimentalMaterialApi::class)
+@Suppress("LongMethod")
```
Don't suppress this warning if you can split a method

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+fun FeedingApproveItem(
+    title: String,
+    status: FeedStatus,
+    feededBy: String,
```
There is no such word `feeded`, please replace it with `fed` and ask the designer to do the same :)

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+    feededBy: String,
+    feededDate: String,
+    imageUrl: String,
+    onApporoveClick: () -> Unit,
```
Misspell

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+        ),
+        onClick = {
+            onApporoveClick()
+            onRejectClick()
```
These 2 callbacks must not be here. There are will be separate buttons for them.

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+                .padding(12.dp),
+            horizontalArrangement = Arrangement.spacedBy(16.dp)
+        ) {
+            Card(
```
Do we really need to wrap `AsyncImage` in `Card`?

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+            Column(
+                modifier = Modifier
+                    .weight(1f)
+                    .fillMaxHeight()
```
`.fillMaxSize()` does the same, doesn't it?

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+                modifier = Modifier
+                    .weight(1f)
+                    .fillMaxHeight()
+                    .padding(vertical = 1.dp),
```
Make it `2.dp`

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+                Row(
+                    modifier = Modifier
+                        .fillMaxWidth()
+                        .padding(vertical = 1.dp),
```
Not needed

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingsScreenUi.kt:
```diff
@@ -1,38 +1,69 @@
+    onBack: () -> Unit,
+    onEvent: (FeedingsScreenEvent) -> Unit,
+) {
+    Column {
```
```suggestion
    Column(modifier = Modifier.fillMaxSize()) {
```

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingsScreenContentUI.kt:
```diff
@@ -0,0 +1,72 @@
+@Composable
+internal fun FeedingsScreenContentUI(
+    feedings: ImmutableList<FeedingItem>,
+    padding: PaddingValues,
```
You probably forgot to remove this

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingsScreenContentUI.kt:
```diff
@@ -0,0 +1,72 @@
+        modifier = Modifier
+            .fillMaxSize()
+            .padding(padding)
+            .padding(horizontal = 32.dp)
```
All paddings are redundant here

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingsScreenEvent.kt:
```diff
@@ -0,0 +1,10 @@
+
+internal sealed interface FeedingsScreenEvent {
+
+    val feedingItem: FeedingItem
```
Don't make this property for the interface, not all events are gonna override it.

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingsScreenUi.kt:
```diff
@@ -1,38 +1,69 @@
+private fun FeedingScreenEmptyPreview() {
+    AnimealTheme {
+        FeedingsScreenUI(
+            feedings = listOf<FeedingItem>().toImmutableList(),
```
```suggestion
            feedings = persistentListOf(),
```

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/model/FeedStatus.kt:
```diff
@@ -0,0 +1,38 @@
+    PENDING_RED(R.string.feed_status_pending, R.drawable.ic_pending_red, CustomColor.StatusRed),
+    PENDING_GREY(R.string.feed_status_pending, R.drawable.ic_pending_grey, CustomColor.StatusGrey),
+    REJECTED(R.string.feed_status_rejected, R.drawable.ic_rejected, CustomColor.StatusRed),
+    OUTDATED(R.string.feed_status_outdated, R.drawable.ic_outdated, CustomColor.StatusRed),
```
This status has a different color on the design

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/model/FeedStatus.kt:
```diff
@@ -0,0 +1,38 @@
+        FeedingStatus.outdated -> FeedStatus.OUTDATED
+    }
+
+private const val ORANGE_TIME = 1_800_000
```
```suggestion
private val orangeTime = 30.minutes.inWholeMilliseconds
```

yuriiq replied:
fixed

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/model/FeedStatus.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+
+private const val ORANGE_TIME = 1_800_000
+private const val RED_TIME = 21_600_000
```
```suggestion
private val redTime = 1.hours.inWholeMilliseconds
```

yuriiq replied:
fixed

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/model/FeedingItem.kt:
```diff
@@ -0,0 +1,62 @@
+    val image: String,
+    val elapsedTime: String,
+) : Parcelable {
+    constructor(feedingPoint: FeedingPoint) : this(
```
This constructor is redundant

yuriiq replied:
removed

Fileek commented on library/api/src/main/java/com/epmedu/animeal/api/feeding/FeedingApiImpl.kt:
```diff
@@ -43,23 +43,22 @@ internal class FeedingApiImpl(
         )
     }
 
-    override suspend fun getApprovedFeedingHistories(feedingPointId: String): ApiResult<SearchFeedingHistoriesQuery.Data> {
```
Don't touch this code. You need to get all feeding histories for feedings screen. Create a separate method for this without any filters.

yuriiq replied:
The "status" parameter is optional, a separate method is not required

Fileek replied:
This method filters feedings by feeding point and you don't need this at all, because the requirement is to get all feedings

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/model/FeedingsViewModel.kt:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private suspend fun collectFeedingPoints() {
+        getFeedingPointsUseCase().collect { feedingPoints ->
```
As soon as you create separate methods for getting all feeding histories and feedings in progress without the filter for feeding point id, you will be able to remove this

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/FeedingItemUI.kt:
```diff
@@ -0,0 +1,157 @@
+                    horizontalArrangement = Arrangement.SpaceBetween
+                ) {
+                    Text(
+                        text = title,
```
To always show date
```suggestion
                        text = title,
                        modifier = Modifier.weight(1f),
```

yuriiq replied:
added

yuriiq committed:
[EPMEDU-1506]: fix notes

Fileek commented on shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/domain/model/FeedingStatus.kt:
```diff
@@ -0,0 +1,17 @@
+}
+fun toFeedingStatus(feedingStatus: type.FeedingStatus): FeedingStatus =
+    when (feedingStatus) {
+        type.FeedingStatus.outdated -> FeedingStatus.OUTDATED
```
You still use data models in domain. Move mapper to data layer please

Fileek commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/di/FeedingsModule.kt:
```diff
@@ -0,0 +1,21 @@
+
+@Module
+@InstallIn(ViewModelComponent::class)
+internal object FeedingsModule {
```
You still separate modules, while you already have `FeedingsDomainModule`

Commit:
Merge branch 'develop' into feature/EPMEDU-1506_View_list_of_feedings

Commit:
Refactor and fixes:

  - Refactor logic to request only feeding points related to existing feedings
  - Add limit for 10 days for feedings queries
  - Fix bug when pending feeding was considered as current feeding in progress

Commit:
Remove navigation from more screen, since it is a part of another story

Commit:
Fix name of GetAllFeedingPointsUseCase file

Commit:
Merge remote-tracking branch 'origin/develop' into feature/EPMEDU-1506_View_list_of_feedings

# Conflicts:
#	shared/feature/feeding/src/main/java/com/epmedu/animeal/feeding/data/repository/FeedingRepositoryImpl.kt

Commit:
Refactor presentation layer of feedings feature

Commit:
Fix detekt

Commit:
Fix outdated color

ozkarram commented on shared/feature/feedings/src/main/java/com/epmedu/animeal/feedings/presentation/viewmodel/FeedingsViewModel.kt:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private fun createFeedingModel(
+        feeding: Feeding,
```
what happens if I use this function with a `Feeding` and a `FeedingPoint` objects non-consistent? There is no check of these being part of the same entity

Fileek replied:
@ozkarram Why should I check it if I get feeding point by `feedingPointId` attached to feeding? It there is no such feeding point, feeding will be filtered out by `mapNotNull`

ozkarram replied:
Yes, specifically where you are calling it that's correct, but if it is called somewhere else it's possible to make a `FeedingModel` from different non-related `Feeding` and `FeedingPoint` objects.

Fileek replied:
Fixed

ozkarram reviewed (changes requested):
Is there a way to fake the moderator role? if not part of the scope of the ticket, add it just for reference.

Fileek commented:
> Is there a way to fake the moderator role? if not part of the scope of the ticket, add it just for reference.

Sure. 

1. Go to admin panel from [this KB page](https://kb.epam.com/pages/viewpage.action?pageId=1712926476) **using the same environment as your app and logging in as an administrator.**
2. Open Users tab.
3. Find your current account and open its details.
4. Press Add to group.
5. Select either Administrator or Moderator.
6. Press Add to group in the dialog.
7. Relaunch the app.

Commit:
Add a check for feeding point id
